### PR TITLE
State finite shots warn

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
Some return types use the state information of the device. These are `qml.state`, `qml.density_matrix`, `qml.vn_entropy` and `qml.mutual_info`.

When setting the device's `shots` attribute to finite shots, these return types will still return analytic results. Therefore, sampling does not affect the outputs.

**Description of the Change:**

Adds a warning when using the above-mentioned return types with `shots!=None` such that users are aware that they are getting analytic results unaffected by sampling.

**Benefits:**
More informative QNode behaviours.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: tests/returntypes/test_new_return_types.py [hunk 1]
```diff
@@ -77,15 +86,19 @@ def circuit(x):
...
+            res = qml.execute_new(tapes=[qnode.tape], device=dev, gradient_fn=None)
 
         assert res[0].shape == (2**d_wires, 2**d_wires)
         assert isinstance(res[0], np.ndarray)
 
     @pytest.mark.parametrize("device", devices)
     def test_expval(self, device, shots):
         """Return a single expval."""
-        dev = qml.device(device, wires=2)
+        dev = qml.device(device, wires=2, shots=shots)
```
Comment (antalszava) [inline tests/returntypes/test_new_return_types.py:101]: Not completely in scope, but stumbled upon these cases that should be parametrized by `shots` too.

**Comment by antalszava** [comment]
Note: the new warnings could be emitted by existing tests. To check that none of the new warnings is emitted, first errors were added prior to the warnings. See commit https://github.com/PennyLaneAI/pennylane/pull/2918/commits/92f52a80fe524176fedae7ccd532ee25574ac0a9 for which the test suite ran successfully.

The subsequent commits changed to having warnings and caught warnings where required (in test cases that have previously been failing).

File: tests/returntypes/test_new_return_types.py (hunk 1 above)
Comment (rmoyard) [reply tests/returntypes/test_new_return_types.py:101]: Thanks!

**Review by rmoyard — APPROVED** [review]
Thank you @antalszava for those warnings! I was wondering if we could raise just one warning with a generic message about analytical values with shots, instead of creating multiple messages. It would be nice not to overload the code here. It is just a personal opinion, happy to approve it otherwise 👍

**Pull request merged**